tests/unit/test_charm_relations.py

```diff
@@ -0,0 +1,170 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import unittest
+from pathlib import Path
+from typing import Tuple
+from unittest.mock import Mock, call, patch
+
+from charm import AUSFOperatorCharm
+from ops import ActiveStatus, testing
+
+from lib.charms.tls_certificates_interface.v3.tls_certificates import ProviderCertificate
+
+TEST_NRF_URL = "https://nrf-example.com:1234"
+TEST_POD_IP = b"1.2.3.4"
+
+
+class TestCharmRelations(unittest.TestCase):
+    def setUp(self):
+        self.namespace = "whatever"
+        self.harness = testing.Harness(AUSFOperatorCharm)
+        self.harness.set_model_name(name=self.namespace)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.set_leader(is_leader=True)
+        self.harness.begin()
+
+    @patch("charm.check_output")
+    @patch(
+        "charms.tls_certificates_interface.v3.tls_certificates.TLSCertificatesRequiresV3.get_assigned_certificates",  # noqa: E501
+    )
+    @patch("ops.model.Container.remove_path")
+    def test_given_charm_is_in_active_state_when_certificates_relation_broken_then_certificate_csr_and_private_key_are_removed(  # noqa: E501
+        self, patched_remove_path, patched_get_assigned_certificates, patched_check_output
+    ):
+        test_private_key = b"whatever private key"
+        test_csr = b"whatever csr"
+        test_certificate = "whatever certificate"
+        self.harness.set_can_connect(container="ausf", val=True)
+        certificates_relation_id, _ = self._create_charm_relations_and_relation_data()
+        self.harness.add_storage(storage_name="certs", attach=True)
+        self.harness.add_storage(storage_name="config", attach=True)
+        patched_check_output.return_value = TEST_POD_IP
+        root = self.harness.get_filesystem_root("ausf")
+        (root / "support/TLS/ausf.csr").write_text(test_csr.decode())
+        (root / "support/TLS/ausf.key").write_text(test_private_key.decode())
+        (root / "support/TLS/ausf.pem").write_text(test_certificate)
+        provider_certificate = Mock(ProviderCertificate)
+        provider_certificate.certificate = test_certificate
+        provider_certificate.csr = test_csr.decode()
+        patched_get_assigned_certificates.return_value = [
+            provider_certificate,
+        ]
+        self.harness.charm.on.update_status.emit()
+        self.harness.evaluate_status()
+        self.assertEqual(self.harness.model.unit.status, ActiveStatus())
+
+        self.harness.remove_relation(certificates_relation_id)
+
+        patched_remove_path.assert_has_calls(
+            [
+                call(path="/support/TLS/ausf.key"),
+                call(path="/support/TLS/ausf.csr"),
+                call(path="/support/TLS/ausf.pem"),
+            ]
```

Comment:
Instead of checking that the calls are made in a given order, with harness we can now check that the file is (not) there.  For instance if you wanted to check the file content:

```
self.assertEqual((root / "vault/certs/ca.pem").read_text(), EXAMPLE_CA)
```

But this can be edited to check that the file isn't there anymore.


This will reduce test flakiness in case the order changes or the approach to place the file there changes.

Reply:
Fixed.